Circular association intermittently fails schema generation

Related to #7 

When using a circular association (in my example via one-to-many association), schema generation will fail ~50% of the time on startup: 

```groovy
class User {
    Address address
    static constraints = {
        address nullable: true
    }
    static graphic = true
}

class Address {
    Set<User> owners
    static hasMany = [
        owners: User
    ]
}
```

results in...

```java
Caused by: graphql.AssertException: type Address not found in schema
	at graphql.Assert.assertNotNull(Assert.java:13)
	at graphql.schema.GraphQLTypeResolvingVisitor.visitGraphQLTypeReference(GraphQLTypeResolvingVisitor.java:42)
	at graphql.schema.GraphQLTypeReference.accept(GraphQLTypeReference.java:45)
	at graphql.schema.TypeTraverser$TraverserDelegateVisitor.enter(TypeTraverser.java:73)
	at graphql.util.Traverser.traverse(Traverser.java:106)
	at graphql.schema.TypeTraverser.doTraverse(TypeTraverser.java:60)
	at graphql.schema.TypeTraverser.depthFirst(TypeTraverser.java:52)
	at graphql.schema.TypeTraverser.depthFirst(TypeTraverser.java:40)
	at graphql.schema.SchemaUtil.replaceTypeReferences(SchemaUtil.java:106)
	at graphql.schema.GraphQLSchema$Builder.build(GraphQLSchema.java:341)
	at graphql.schema.GraphQLSchema$Builder.build(GraphQLSchema.java:314)
	at org.grails.gorm.graphql.Schema.generate(Schema.groovy:446)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
```

I've include a full project reproducing this at https://github.com/jamesdh/gorm-graphql-bug. Either running the app or running integration tests should reproduce it ~50% of the time.

--- a/core/src/main/groovy/org/grails/gorm/graphql/Schema.groovy
+++ b/core/src/main/groovy/org/grails/gorm/graphql/Schema.groovy
@@ -253,6 +253,8 @@ class Schema {
                         childrenNotMapped.add(entity)
                     }
                     continue
+                } else if (!mapping.operations.all.enabled) {
+                    continue
                 }
 
                 List<GraphQLFieldDefinition.Builder> queryFields = []
@@ -264,8 +266,11 @@ class Schema {
                 List<Closure> postIdentityExecutables = []
                 InterceptorInvoker queryInterceptorInvoker = new QueryInterceptorInvoker()
 
+                ProvidedOperation queryOperation = mapping.operations.query
+                ProvidedOperation mutationOperation = mapping.operations.mutation
+
                 ProvidedOperation getOperation = mapping.operations.get
-                if (getOperation.enabled) {
+                if (queryOperation.enabled && getOperation.enabled) {
 
                     DataFetcher getFetcher = dataFetcherManager.getReadingFetcher(entity, GET).orElse(new SingleEntityDataFetcher(entity))
 
@@ -288,7 +293,7 @@ class Schema {
                 }
 
                 ListOperation listOperation = mapping.operations.list
-                if (listOperation.enabled) {
+                if (queryOperation.enabled && listOperation.enabled) {
 
                     DataFetcher listFetcher = dataFetcherManager.getReadingFetcher(entity, LIST).orElse(null)
 
@@ -332,7 +337,7 @@ class Schema {
                 }
 
                 ProvidedOperation countOperation = mapping.operations.count
-                if (countOperation.enabled) {
+                if (queryOperation.enabled && countOperation.enabled) {
 
                     DataFetcher countFetcher = dataFetcherManager.getReadingFetcher(entity, COUNT).orElse(new CountEntityDataFetcher(entity))
 
@@ -358,7 +363,7 @@ class Schema {
                 GraphQLDataBinder dataBinder = dataBinderManager.getDataBinder(entity.javaClass)
 
                 ProvidedOperation createOperation = mapping.operations.create
-                if (createOperation.enabled && !Modifier.isAbstract(entity.javaClass.modifiers)) {
+                if (mutationOperation.enabled && createOperation.enabled && !Modifier.isAbstract(entity.javaClass.modifiers)) {
                     if (dataBinder == null) {
                         throw new DataBinderNotFoundException(entity)
                     }
@@ -388,7 +393,7 @@ class Schema {
                 }
 
                 ProvidedOperation updateOperation = mapping.operations.update
-                if (updateOperation.enabled) {
+                if (mutationOperation.enabled && updateOperation.enabled) {
                     if (dataBinder == null) {
                         throw new DataBinderNotFoundException(entity)
                     }
@@ -422,7 +427,7 @@ class Schema {
                 }
 
                 ProvidedOperation deleteOperation = mapping.operations.delete
-                if (deleteOperation.enabled) {
+                if (mutationOperation.enabled && deleteOperation.enabled) {
 
                     DeletingGormDataFetcher deleteFetcher = dataFetcherManager.getDeletingFetcher(entity).orElse(new DeleteEntityDataFetcher(entity))
 

--- a/core/src/main/groovy/org/grails/gorm/graphql/entity/dsl/Operations.groovy
+++ b/core/src/main/groovy/org/grails/gorm/graphql/entity/dsl/Operations.groovy
@@ -14,6 +14,9 @@ import org.grails.gorm.graphql.entity.operations.ProvidedOperation
 @CompileStatic
 class Operations {
 
+    ProvidedOperation mutation = new ProvidedOperation()
+    ProvidedOperation query = new ProvidedOperation()
+    ProvidedOperation all = new ProvidedOperation()
     ProvidedOperation get = new ProvidedOperation()
     ListOperation list = new ListOperation()
     ProvidedOperation create = new ProvidedOperation()

--- a/core/src/test/groovy/org/grails/gorm/graphql/DisableAllOpSpec.groovy
+++ b/core/src/test/groovy/org/grails/gorm/graphql/DisableAllOpSpec.groovy
@@ -1,0 +1,35 @@
+package org.grails.gorm.graphql
+
+import grails.gorm.annotation.Entity
+import graphql.schema.GraphQLSchema
+import org.grails.datastore.gorm.GormEntity
+import org.grails.gorm.graphql.entity.dsl.GraphQLMapping
+
+class DisableAllOpSpec extends HibernateSpec {
+
+    @Override
+    List<Class> getDomainClasses() {
+        [AllDisabledOpEntity]
+    }
+
+    void "test that disable all operation in clean way"() {
+
+        when:
+        GraphQLSchema schema = new Schema(hibernateDatastore.mappingContext)
+                .generate()
+
+        then:
+        schema.queryType.fieldDefinitions.isEmpty()
+        schema.mutationType.fieldDefinitions.isEmpty()
+    }
+}
+
+@Entity
+class AllDisabledOpEntity implements GormEntity<AllDisabledOpEntity> {
+
+    String prop
+
+    static graphql = GraphQLMapping.build {
+        operations.all.enabled false
+    }
+}

--- a/core/src/test/groovy/org/grails/gorm/graphql/DisableReadOpSpec.groovy
+++ b/core/src/test/groovy/org/grails/gorm/graphql/DisableReadOpSpec.groovy
@@ -1,0 +1,40 @@
+package org.grails.gorm.graphql
+
+import grails.gorm.annotation.Entity
+import graphql.schema.GraphQLSchema
+import org.grails.datastore.gorm.GormEntity
+import org.grails.gorm.graphql.entity.dsl.GraphQLMapping
+
+class DisableReadOpSpec extends HibernateSpec {
+
+    @Override
+    List<Class> getDomainClasses() {
+        [ReadDisabledEntity]
+    }
+
+    void "test that disable all operation in clean way"() {
+
+        when:
+        GraphQLSchema schema = new Schema(hibernateDatastore.mappingContext)
+                .generate()
+
+        then:
+        schema.queryType.fieldDefinitions.isEmpty()
+
+        and:
+        schema.mutationType.fieldDefinitions.size() == 3
+        schema.mutationType.fieldDefinitions.find {it.name == "readDisabledEntityCreate"}
+        schema.mutationType.fieldDefinitions.find {it.name == "readDisabledEntityDelete"}
+        schema.mutationType.fieldDefinitions.find {it.name == "readDisabledEntityUpdate"}
+    }
+}
+
+@Entity
+class ReadDisabledEntity implements GormEntity<ReadDisabledEntity> {
+
+    String prop
+
+    static graphql = GraphQLMapping.build {
+        operations.query.enabled false
+    }
+}

--- a/core/src/test/groovy/org/grails/gorm/graphql/ReadOnlyOpSpec.groovy
+++ b/core/src/test/groovy/org/grails/gorm/graphql/ReadOnlyOpSpec.groovy
@@ -1,0 +1,39 @@
+package org.grails.gorm.graphql
+
+import grails.gorm.annotation.Entity
+import graphql.schema.GraphQLSchema
+import org.grails.datastore.gorm.GormEntity
+import org.grails.gorm.graphql.entity.dsl.GraphQLMapping
+
+class ReadOnlyOperationsSpec extends HibernateSpec {
+
+    @Override
+    List<Class> getDomainClasses() {
+        [ReadOpOnlyEntity]
+    }
+
+    void "test that only read operation are enabled in clean way"() {
+
+        when:
+        GraphQLSchema schema = new Schema(hibernateDatastore.mappingContext)
+                .generate()
+
+        then:
+        schema.mutationType.fieldDefinitions.isEmpty()
+
+        and:
+        schema.queryType.fieldDefinitions.size() == 3
+        schema.queryType.fieldDefinitions.find {it.name == "readOpOnlyEntity"}
+        schema.queryType.fieldDefinitions.find {it.name == "readOpOnlyEntityList"}
+        schema.queryType.fieldDefinitions.find {it.name == "readOpOnlyEntityCount"}
+    }
+}
+
+@Entity
+class ReadOpOnlyEntity implements GormEntity<ReadOpOnlyEntity> {
+    String prop
+
+    static graphql = GraphQLMapping.build {
+        operations.mutation.enabled false
+    }
+}

--- a/docs/src/main/docs/guide/customizations/operations.adoc
+++ b/docs/src/main/docs/guide/customizations/operations.adoc
@@ -1,8 +1,17 @@
 == Provided Operations
 
-The 2 query (get, list) and 3 mutation (create, update, delete) operations provided by default for all entities mapped with GraphQL are called the provided operations.
+The 3 query (get, list, count) and 3 mutation (create, update, delete) operations provided by default for all entities mapped with GraphQL are called the provided operations.
 
 It is possible to customize the provided operations as well as disable their creation.
+
+=== Disabling All Operations
+
+[source,groovy]
+----
+static graphql = GraphQLMapping.build {
+    operations.all.enabled false
+}
+----
 
 === Disabling A Provided Operation
 
@@ -19,6 +28,25 @@ static graphql = GraphQLMapping.build {
     //operations.delete
 }
 ----
+
+=== Disabling All Mutation(Write) Operation
+
+[source,groovy]
+----
+static graphql = GraphQLMapping.build {
+    operations.mutation.enabled false
+}
+----
+
+=== Disabling All Query(Read) Operation
+
+[source,groovy]
+----
+static graphql = GraphQLMapping.build {
+    operations.query.enabled false
+}
+----
+
 
 === Metadata
 

--- a/examples/grails-test-app/grails-app/domain/grails/test/app/manyToMany/Classes.groovy
+++ b/examples/grails-test-app/grails-app/domain/grails/test/app/manyToMany/Classes.groovy
@@ -9,11 +9,6 @@ class Classes {
     static hasMany = [students: Student]
 
     static graphQL = GraphQLMapping.build {
-        operations.get.enabled false
-        operations.list.enabled false
-        operations.count.enabled false
-        operations.create.enabled false
-        operations.delete.enabled false
-        operations.update.enabled false
+        operations.all.enabled false
     }
 }

--- a/examples/grails-test-app/grails-app/domain/grails/test/app/manyToMany/Classes.groovy
+++ b/examples/grails-test-app/grails-app/domain/grails/test/app/manyToMany/Classes.groovy
@@ -1,0 +1,19 @@
+package grails.test.app.manyToMany
+
+import org.grails.gorm.graphql.entity.dsl.GraphQLMapping
+
+class Classes {
+    String title
+    String description
+
+    static hasMany = [students: Student]
+
+    static graphQL = GraphQLMapping.build {
+        operations.get.enabled false
+        operations.list.enabled false
+        operations.count.enabled false
+        operations.create.enabled false
+        operations.delete.enabled false
+        operations.update.enabled false
+    }
+}

--- a/examples/grails-test-app/grails-app/domain/grails/test/app/manyToMany/Student.groovy
+++ b/examples/grails-test-app/grails-app/domain/grails/test/app/manyToMany/Student.groovy
@@ -1,0 +1,9 @@
+package grails.test.app.manyToMany
+
+class Student {
+    String name
+
+    static hasMany = [classes: Classes]
+
+    static graphql = true
+}

--- a/examples/grails-test-app/src/test/groovy/grails/test/app/manyToMany/StudentSchemaSpec.groovy
+++ b/examples/grails-test-app/src/test/groovy/grails/test/app/manyToMany/StudentSchemaSpec.groovy
@@ -1,0 +1,22 @@
+package grails.test.app.manyToMany
+
+import grails.test.hibernate.HibernateSpec
+import org.grails.gorm.graphql.Schema
+
+class StudentSchemaSpec extends HibernateSpec {
+
+    @Override
+    List<Class> getDomainClasses() {
+        [Classes, Student]
+    }
+
+    void "test schema generation for many-to-many relationship with on side disabled for all operations"() {
+
+        when:
+        new Schema(hibernateDatastore.mappingContext)
+                .generate()
+
+        then:
+        noExceptionThrown()
+    }
+}

